### PR TITLE
Added information about unofficial docker container

### PR DESCRIPTION
**Pull request by ergoz**
Container tested. Built with libconfig, libwrap and libcap
And it works with Telegram :)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,6 +20,25 @@ serve SSH and HTTPS on the same port.
...
 ===================
 
+Docker container
+----------------
+
+For easy installation and usage you can use unofficial docker container.
+Container built with `libconfig`, `libwrap` and `libcap` support. 
+It can be found on [Docker Hub](https://hub.docker.com/r/riftbit/sslh/)
+
+**Example with --net host**
```
Comment (yrutschle) [inline README.md]: Can you add a description of what "--net host" means? As a non-docker user, I have no idea what the difference is between this setting and the next "exposed ports" setting (I guess the latter means it opens ports on the outside).

File: README.md [hunk 2]
```diff
@@ -20,6 +20,25 @@ serve SSH and HTTPS on the same port.
...
+
+```
+docker run --name sslh -d --env SSLH_OPTS='-p 0.0.0.0:443 --anyprot 192.168.0.1:8443' --net host --privileged --restart always riftbit/sslh
+```
+
+**Example with exposed ports**
+
```
Comment (yrutschle) [inline README.md]: also, how does this interact with an apache and sshd?do these need to be in dockers themselves? do you just need to move  apache off 443? do you need to set sshd to 8443?

**Comment by yrutschle** [comment]
Also (again) how does this work with telegram? just using the 'anyprot' default?

**Comment by iShift** [comment]
Link for docker container without dockerfile of it - bad idea. We don’t know what’s in container

File: README.md (hunk 1 above)
Comment (ergoz) [reply README.md]: It have difference in network type. if set `--net host` - container will work with host network interface directly, like you run application in host system, and all ports exposed. If this option not setted container will work with bridged network interface, and you will need to expose all ports and make additional configurations to allow container work with other addresses.

File: README.md (hunk 2 above)
Comment (ergoz) [reply README.md]: no, apache or sshd can be installed as you wish (in container or not). If you run apache in host system, you can use container gateway ip, for example you run sslh in container with ip 172.17.0.2, apache installed in main system and have port 8443 - you should set 172.17.0.1:8443 in sslh config for apache. sorry for my bad english, i hope you understand me :)

**Comment by ergoz** [comment]
@yrutschle With telegram it works through `anyprot` option and works great. In Russia it is very actual problem ;) And i can try to make documentation more informative with more examples if you interested.

@iShift you can check container at https://microbadger.com/images/riftbit/sslh (but sometimes it lags)
You can see dockerfile at dockerhub only if docker hub used for building container (auto builds mode). But i builds container locally on my own CI/CD like in most my projects and private repositories. :) I

For this container I uses docker build steps. First step (only for binary compillation) i use `debian:stretch`, and for second step (for main usage) i use `debian:stretch-slim` to minimize container size. 

For this container i made build argument to make both version of sslh (you can see it in tags section on docker hub) for fork and select modes.

**Comment by iShift** [comment]
> @iShift you can check container at https://microbadger.com/images/riftbit/sslh (but sometimes it lags)
You can see dockerfile at dockerhub only if docker hub used for building container (automated builds mode). But i builds container locally on my own CI/CD like in most my projects and private repositories. :)

@ergoz if you have Dockerfile - publish it, or make PR for this repo, docker file is most interesting part of that work =)

**Comment by yrutschle** [comment]
Following (this news)[https://techcrunch.com/2018/06/15/tainted-crypto-mining-containers-pulled-from-docker-hub/] which confirms a nagging suspicion I had about Docker distribution, I second @iShift: I'm happy to accept a Dockerfile but will not encourage people to blindly download binaries from the Internet, sorry.

**Comment by guusdk** [comment]
Using Docker (or any container for that matter), is it possible to have transparent proxy support (which allows the services behind sslh to see the original/external IP and port)?

**Comment by ergoz** [comment]
Yes, it is possible. But you need to configure iptables for this.

вт, 30 окт. 2018 г., 12:55 Guus der Kinderen <notifications@github.com>:

**Comment by guusdk** [comment]
I have read that section in the readme of this project over and over (and have it working on one host), but I must admit that I don't understand how it works.

Do you need the same iptables configuration for this to work in a Docker instance? My concern is that the configuration appears to be applicable to a "one-host" setup. How does that translate to Docker (where multiple containers are communicating with each-other, over, presumably, TCP/IP)?

**Comment by jribal** [comment]
Hi, I'm also interested in transparent proxy to allow fail2ban usage.

Like @guusdk I was able to setup sslh with transparent on host. Then I had success with a docker image with transparent using --net host and ssh on the same system. I used the official transparent proxy documentation.

Then I failed making sslh container with --transparent forward to another container no matter if I use --net host or not.
My guess was that the setup should be similar to forwarding to another machine but I got no success using various tutorial, none worked for me.

I would be nice to have a working transparent even inside docker network. I'm willing to help for tests but iptables and routing are really hard for me.

**Comment by eigood** [comment]
> @yrutschle With telegram (MTProto Proxy) it works through `anyprot` option and works great. In Russia it is very actual problem ;) And i can try to make documentation more informative with more examples if you interested.
> 
> @iShift you can check container at https://microbadger.com/images/riftbit/sslh (but sometimes it lags)
> You can see dockerfile at dockerhub only if docker hub used for building container (automated builds mode). But i builds container locally on my own CI/CD like in most my projects and private repositories. :) I
> 
> For this container I uses docker build steps. First step (only for binary compillation) i use `debian:stretch`, and for second step (for main usage) i use `debian:stretch-slim` to minimize container size.
> 
> For this container i made build argument to make both version of sslh (you can see it in tags section on docker hub) for fork and select modes.

I looked at the linked image, and the docker layers it provides; that's not a Dockerfile.  It copies some random file to /sslh, and there is no way for anyone to know how it was compiled.  The same goes for the entrypoint.

Please show us the build system.  My suspicion is that this is a multipler-FROM dockerfile pattern.

**Comment by pataquets** [comment]
Although a public Docker container is very desirable, I would prefer an [Automated Build](https://docs.docker.com/docker-hub/builds/) straight from source.
This way, project gains automatic builds on each commit (no lagging begind of images) and image is auditable and verifiable.
That is, once a Dockerfile is commited in the repo.

**Comment by yrutschle** [comment]
Following this merge 82b8ba547e57ec3dc92af666a40bf61879e57383, I'm closing this.

**Pull request closed without merging**